funsor/contract.py

```diff
@@ -0,0 +1,242 @@
+    return tuple(constants), tuple(new_operands)
+
+
+def _order_lhss(lhs, reduced_vars):
+    assert isinstance(lhs, Finitary)
+
+    components = _partition(lhs.operands, reduced_vars)
+    root_lhs = Finitary(ops.mul, tuple(components[0][0]))
+    if len(components) > 1:
+        remaining_lhs = Finitary(ops.mul, tuple(t for c in components[1:] for t in c[0]))
+    else:
+        remaining_lhs = None
+
+    return root_lhs, remaining_lhs
+
+
+def _simplify_contract(lhs, rhs, reduced_vars):
+    """
+    Reduce free variables that do not appear explicitly in the lhs
+    """
+    meas_vars = frozenset(lhs.inputs)
+    int_vars = frozenset(rhs.inputs)
+    assert reduced_vars <= meas_vars | int_vars
+    progress = False
+    if not reduced_vars <= meas_vars:
+        rhs = rhs.reduce(ops.add, reduced_vars - meas_vars)
+        reduced_vars = reduced_vars & meas_vars
+        progress = True
+    if not reduced_vars <= int_vars:
+        lhs = lhs.reduce(ops.add, reduced_vars - int_vars)
+        reduced_vars = reduced_vars & int_vars
+        progress = True
+
+    if progress:
+        return Contract(lhs, rhs, reduced_vars)
+
+    return None
+
+
+class Contract(Funsor):
+
+    def __init__(self, lhs, rhs, reduced_vars):
+        assert isinstance(lhs, Funsor)
+        assert isinstance(rhs, Funsor)
+        assert isinstance(reduced_vars, frozenset)
+        inputs = OrderedDict([(k, d) for t in (lhs, rhs)
+                              for k, d in t.inputs.items() if k not in reduced_vars])
+        output = rhs.output
+        super(Contract, self).__init__(inputs, output)
+        self.lhs = lhs
+        self.rhs = rhs
+        self.reduced_vars = reduced_vars
+
+    def eager_subs(self, subs):
+        # basically copied from Reduce.eager_subs
+        subs = tuple((k, v) for k, v in subs if k not in self.reduced_vars)
+        if not any(k in self.inputs for k, v in subs):
+            return self
+        if not all(self.reduced_vars.isdisjoint(v.inputs) for k, v in subs):
+            raise NotImplementedError('TODO alpha-convert to avoid conflict')
+        return Contract(self.lhs.eager_subs(subs), self.rhs.eager_subs(subs),
+                        self.reduced_vars)
+
+
+@optimize.register(Contract, ATOMS[1:], ATOMS[1:], frozenset)
+@eager.register(Contract, ATOMS[1:], ATOMS[1:], frozenset)
+def contract_ground_ground(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    return (lhs * rhs).reduce(ops.add, reduced_vars)
+
+
+@eager.register(Contract, Tensor, Tensor, frozenset)
+def eager_contract_tensor_tensor(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    if reduced_vars:
+        out_vars = (frozenset(lhs.inputs) | frozenset(rhs.inputs)) - reduced_vars
+    else:
+        out_vars = {}
+
+    out_inputs = OrderedDict([
```

Comment:
This is nondeterministic (the ordering depends on frozenset details). Elsewhere in Funsor I've tried to preserve the ordering of inputs. Would you mind changing the computation here to something like:
```py
otu_inputs = OrderedDict((k, d) for t in (lhs, rhs)
                         for k, d in t.inputs.items() if k not in reduced_vars)
out_vars = frozenset(inputs)
```